rt4914/KlinicCon-Backend
# Fixes #8 feature request: Add Devise Gem with Configurable

1. Updated the code base to have only rails api 
2. Only Devise gem is added with below customization
3. Two controllers registrations & sessions controller is added along with enabling the confirmable (This would send or prompt for confirming the mail address)
4. One bug thats existing in devise for sending fake session to the request headers is provided under concerns (rack_session_fix.rb)
5. cors.rb is uncommented to ensur cross origin is enabled so that external application is provided access
6. In user.rb model, enabled the confirmable 
7. In routes.rb overridden the paths of sessions and registrations
8. Created migration file to enable confirmable
9. Uncommented in schema.rb fields related to confirmable
10. Added in gem file, jsonapi-serializer this is needed to convert or serialize ruby object to JSON
11. User Serializer is added and only id, email and created_at is provided as part of user object
12. Registrations and sessions controller is updated with response_to and response_with so that json is returned (as its rails only api, we are not making use of views and only json response is returned with right status code and message)
13. JSON response messages are modified to have below structure
### Response

> { code: 200, message: 'Signed up successfully', data: { id: 5, email: 'xyz@mail.com', created_at: '2024-08-24T08:00:00.7898"}

Resolves #8

### Description
Summary: Addition & configuration of devise gem
Dependencies: Devise, jsonapi-serializer

### Type of change



### How to test this PR?
Once configuration is completed, below api's can be verified

1. Signup, use below api for signingup/registration, below payload shall be used for sending request and response is displayed below

![CleanShot 2024-08-24 at 10 53 13](https://github.com/user-attachments/assets/eb2f4c66-c340-46c0-bbc0-f62792bb7f4d)


2. Run the mailcatcher so that configured mail address in devise.rb should receive confirmation mail as shown below

**- Run mailcatcher using this command on terminal:** `mailcatcher`

It should run like below:
```
Starting MailCatcher v0.10.0
==> smtp://127.0.0.1:1025
==> http://127.0.0.1:1080
```

![CleanShot 2024-08-23 at 23 04 40](https://github.com/user-attachments/assets/8b371a7b-7a34-45fa-851d-572b80d9beb3)

3. Confirm the mail by clicking on the link

![CleanShot 2024-08-23 at 23 05 56](https://github.com/user-attachments/assets/bd9e7ca9-a68e-43bc-93e9-26dffc904440)

4. Once link is clicked, users account would be marked as confirmed and we are good to login 
5. Use below api for login/signin, (expected response should be successful login as provided in the sessions controller)

![CleanShot 2024-08-24 at 10 54 17](https://github.com/user-attachments/assets/2d87717f-ca63-41b6-bf6c-f25556e6e648)

6. Verify for registration by providing same email id, then we would get error as shown below 

![CleanShot 2024-08-24 at 10 54 43](https://github.com/user-attachments/assets/8252ed78-ad1c-4cf9-83aa-effd6df58729)
